CARDS-2154: cards-utils tests

sofi2002sofi opened the pull request.

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

tests for DenyScriptsSlingPostProcessor

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

tests for ContentTypeSetter and StatusCodeSetter

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

tests for ResourceToCSVAdapterFactory

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

tests for ResourceToJsonAdapterFactory, ResourceToMarkdownAdspterFactory, ResourceToTextAdapterFactory

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

tests for BareProcessor

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

tests for DeepProcessor, DereferenceProcessor, IdentificationProcessor, PropertiesProcessor, SimpleProcessor

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/ResourceToCSVAdapterFactoryTest.java:
```diff
@@ -0,0 +1,94 @@
+ * @version $Id $
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class ResourceToCSVAdapterFactoryTest
```
More tests:

- `getAdapterWithNoProcessorsReturnsResourcePath` - has an empty list of processors
- `getAdapterUsesFirstProcessorThatCanProcess` - has 3 processors, first cannot process, check second's output is used, verify first's and third's `serialize` is never invoked

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/ResourceToCSVAdapterFactoryTest.java:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    @Test
+    public void getAdapterForResourceAdaptableObjectReturnsResourcePath()
```
```suggestion
    public void getAdapterForUnsupportedResourceReturnsResourcePath()
```

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/ResourceToJsonAdapterFactoryTest.java:
```diff
@@ -0,0 +1,166 @@
+
+        when(processor.isEnabledByDefault(adaptable)).thenReturn(true);
+        when(processor.getName()).thenReturn("name");
+        when(processor.canProcess(adaptable)).thenReturn(true);
```
These lines are the same in all 3 tests, they should be moved in a `mockWorkingProcessor` method.

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/ResourceToJsonAdapterFactoryTest.java:
```diff
@@ -0,0 +1,166 @@
+    {
+        Resource adaptable = mock(Resource.class);
+        ResourceMetadata resourceMetadata = mock(ResourceMetadata.class);
+        when(adaptable.getResourceMetadata()).thenReturn(resourceMetadata);
```
This code seems useless to me; the fact that it is actually needed here suggests that the actual code can be improved. Try moving the `adaptTo(Node.class)` line earlier in the method, and add a check for null.

Normally, in test-driven development, the tests are supposed to influence the code, not the other way around, so you should feel free to also change the code being tested when you feel that it can be improved, not just write tests that perfectly match the existing code.

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/ResourceToJsonAdapterFactoryTest.java:
```diff
@@ -0,0 +1,166 @@
+ * @version $Id $
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class ResourceToJsonAdapterFactoryTest
```
More tests:
- `getAdapterUsesAllSupportedAndEnabledProcessors` - receives a list of processors that either support or not the resource, and that are either enabled or disabled by default
- `getAdapterSortsProcessors` - receives a list of processors in the wrong order; to make the order validation easier to check, write a simple class that receives a priority and 2 integers `a` and `b` in its constructor, and implements `ResourceJsonProcessor.processProperty` as `return Json.createValue(input == null ? b : ((JsonNumber) input).intValue() * a + b);`; instantiate 4 such processors with different prime values for `a` and `b`, and serialize a number; make sure the output matches the expected result
- `getAdapterWithNoProcessorsReturnsEmptyJsonObject` - uses an empty list of processors
- `getAdapterWithNoSupportedProcessorsReturnsEmptyJsonObject` - uses one processor that cannot process the resource
- `getAdapterWithRecursiveReferencesUsesResourcePathForNestedReferences`
- `getAdapterUsesResourceSelectorsToDisableDefaultProcessors`
- `getAdapterUsesResourceSelectorsToEnableProcessors`
- `getAdapterWithBothEnableAndDisableSelectorsPrioritizesEnable`

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/ResourceToMarkdownAdapterFactoryTest.java:
```diff
@@ -0,0 +1,96 @@
+ * @version $Id $
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class ResourceToMarkdownAdapterFactoryTest
```
This is very similar to `ResourceToCSVAdapterFactoryTest`, so those comments apply here too.

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/ResourceToTextAdapterFactoryTest.java:
```diff
@@ -0,0 +1,96 @@
+ * @version $Id $
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class ResourceToTextAdapterFactoryTest
```
This is very similar to `ResourceToCSVAdapterFactoryTest`, so those comments apply here too.

The `ResourceToTextAdapterFactory` is a bit special, so only this test class needs a special test:
- `getAdapterForJcrPropertyResourceReturnsNull`.

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/ResourceToTextAdapterFactoryTest.java:
```diff
@@ -0,0 +1,96 @@
+        assertEquals(TEST_FORM_PATH, adapter);
+    }
+
+}
```
Although tiny and already "covered" by other tests, you should write two tests for `CSVString` as well:
- `toStringReturnsInputData`
- `toStringWithNullDataReturnsNull`

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/utils/internal/DenyScriptsSlingPostProcessorTest.java:
```diff
@@ -0,0 +1,149 @@
+ * @version $Id$
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class DenyScriptsSlingPostProcessorTest
```
More tests:
- `processAllowsNullResource` - `getResource` returns `null`
- `processScriptResourceIgnoresCase` - test with content type `application/TypeScript`
- `processHtmlResourceIgnoresCase` - test with content type `application/XHTML`

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/utils/internal/DenyScriptsSlingPostProcessorTest.java:
```diff
@@ -0,0 +1,149 @@
+    private DenyScriptsSlingPostProcessor denyScriptsSlingPostProcessor;
+
+    @Test
+    public void processForNullResourceMetadataInModification() throws Exception
```
```suggestion
    public void processAllowsResourceWithNullResourceMetadata() throws Exception
```

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/utils/internal/DenyScriptsSlingPostProcessorTest.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Test
+    public void processForNullContentTypeMetadataResourceInModification() throws Exception
```
```suggestion
    public void processAllowsResourceWithNullContentType() throws Exception
```

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/utils/internal/DenyScriptsSlingPostProcessorTest.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Test
+    public void processForScriptContentTypeMetadataResourceInModificationThrowsException()
```
```suggestion
    public void processScriptResourceThrowsException()
```

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/utils/internal/DenyScriptsSlingPostProcessorTest.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Test
+    public void processForHtmlContentTypeMetadataResourceInModificationThrowsException()
```
```suggestion
    public void processHtmlResourceThrowsException()
```

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/utils/internal/DenyScriptsSlingPostProcessorTest.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Test
+    public void processForContentTypeMetadataResourceInModification() throws Exception
```
```suggestion
    public void processAllowsOtherContentTypeResource() throws Exception
```

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/utils/internal/DenyScriptsSlingPostProcessorTest.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Test
+    public void processForAdminUser() throws Exception
```
```suggestion
    public void processAllowsAllForAdminUser() throws Exception
```

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/utils/internal/DenyScriptsSlingPostProcessorTest.java:
```diff
@@ -0,0 +1,149 @@
+        SlingHttpServletRequestImpl request = mock(SlingHttpServletRequestImpl.class);
+        when(request.getRemoteUser()).thenReturn("admin");
+        this.denyScriptsSlingPostProcessor.process(request, List.of());
+        verify(request, times(0)).getResourceResolver();
```
This test isn't specific enough. Just looking at this test code isn't enough to inform me whether the code passes the test because there are no changes, or because the admin is allowed to perform all requested changes. To be thorough, this should set up a "here's a script file being processed" scenario, just like the failing test for scripts above, but the validation is different: there is no exception, and also there's no interaction with the resolver.

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/internal/BareProcessorTest.java:
```diff
@@ -0,0 +1,330 @@
+    }
+
+    @Test
+    public void leaveSerializeCreatedAndLastModifiedAndContent() throws RepositoryException, ParseException
```
```suggestion
    public void leaveSerializesCreatedAndLastModifiedAndFileContent() throws RepositoryException, ParseException
```

But I would split this into two tests, one for `created` and `lastModified`, and one for file content.

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/internal/BareProcessorTest.java:
```diff
@@ -0,0 +1,330 @@
+
+        assertNotNull(jsonObject);
+        assertTrue(jsonObject.containsKey("created"));
+        assertEquals(date.getTime(), format.parse(created.substring(1, created.length() - 1)));
```
This looks too complicated, which means that something could be improved. Did you try using `jsonObject.getString` instead of `get(..).toString`?

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/internal/BareProcessorTest.java:
```diff
@@ -0,0 +1,330 @@
+    public void leaveSerializeCreatedAndLastModifiedAndContent() throws RepositoryException, ParseException
+    {
+        Node node = mock(Node.class);
+        when(this.depth.get()).thenReturn(1, 0);
```
Another test that can be written: `leaveWithNonRootNodeDoesNotAddMetadata`, which uses `2, 1` here and checks that the metadata wasn't added.

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/ResourceToCSVAdapterFactoryTest.java:
```diff
@@ -0,0 +1,94 @@
+/**
+ * Unit tests for {@link ResourceToCSVAdapterFactory}.
+ *
+ * @version $Id $
```
This doesn't work with the extra space, it must be `$Id$`. Same for all the other files where a space is present.

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/internal/DereferenceProcessorTest.java:
```diff
@@ -0,0 +1,292 @@
+    }
+
+    @Test
+    public void isEnabledByDefaultTest()
```
Oh, I think each processor needs this test, even if the method is not overridden in a class, just to check that it is indeed false.

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/internal/DereferenceProcessorTest.java:
```diff
@@ -0,0 +1,292 @@
+        Property property = node.getProperty("jcr:baseVersion");
+        JsonValue json = Json.createValue(FORM_TYPE);
+        JsonValue jsonValue = this.dereferenceProcessor.processProperty(node, property, json, this::serializeNode);
+        assertNotNull(jsonValue);
```
This needs an `assertEquals` check as well.

```
        assertEquals(property.getNode().getPath(), ((JsonString) jsonValue).getString());
```

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/internal/IdentificationProcessorTest.java:
```diff
@@ -0,0 +1,153 @@
+        assertTrue(jsonObject.containsKey("@path"));
+        assertEquals(Json.createValue(TEST_FORM_PATH), jsonObject.getJsonString("@path"));
+        assertTrue(jsonObject.containsKey("@name"));
+        assertEquals(Json.createValue("f1"), jsonObject.getJsonString("@name"));
```
Why not the simpler:

```
        assertEquals("f1", jsonObject.getString("@name"));
```

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/internal/IdentificationProcessorTest.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    @Test
+    public void leaveAddsPathAndNameAndReferencedParameters() throws RepositoryException
```
I would split this into 2 methods:

- `leaveAddsPathAndNameParameters`
- `leaveAddsReferencedParameter`

Also, I would test the reference for 2 different nodes, the form (for `false`) and the questionnaire (for `true`).

sdumitriu commented on modules/utils/src/test/java/io/uhndata/cards/serialize/internal/SimpleProcessorTest.java:
```diff
@@ -0,0 +1,110 @@
+ * @version $Id $
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class SimpleProcessorTest
```
This is missing a positive test: `processPropertyReturnsInput`.

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

fixes for ResourceToCSVAdapterFactory

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

fixes for ResourceToMarkdownAdapterFactory

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

fixes for ResourceToTextAdapterFactory

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

tests for CSVString

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

fixes for DenyScriptsSlingPostProcessorTest

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

fixes for BareProcessorTest

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

fixes for DeepProcessorTest, DeferenceProcessorTest, PropertiesProcessorTest, PropertiesProcessorTest

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

fixes for IdentificationProcessorTest

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

fixes for SimpleProcessorTest

sofi2002sofi committed:
CARDS-2154: Increase test coverage cards-utils

fixes for ResourceToJsonAdapterFactoryTest

veronikaslc commented:
`ResourceToJsonAdapterFactoryTest`: Lines 267-208 not covered from `ResourceToJsonAdapterFactory`
`BareProcessorTest` not covered cases of lines 141 and 206-209 of `BareProcessor`
`DereferenceProcessorTest`: not covered line 127 of `DereferenceProcessor`
`SimpleProcessorTest`: not covered line 101 of `SimpleProcessor`

veronikaslc commented:
[jacoco.zip](https://github.com/data-team-uhn/cards/files/11776256/jacoco.zip)
`jacoco:report` html file attached

sofi2002sofi committed:
added new test cases for ResourceToJsonAdapterFactory, BareProcessor, SimpleProcessor, and DereferenceProcessor

veronikaslc committed:
CARDS-2154: Increase test coverage cards-utils
Adjusted test coverage wrt current codeabse volume
Removed 2 outdated tests due to CARDS-2353: Move the @reference JSON processor in its own module and disable it by default
Adjusted some outdated tests
Added missing pom dependencies

sofi2002sofi committed:
CARDS-2154: Increase test coverage of cards-utils

sdumitriu committed:
CARDS-2154: Increase test coverage of cards-utils
Modernize the revived tests and complete the module coverage to 100%:
- migrate the tests to the current platform APIs: jakarta.json,
  SlingJakartaHttpServletRequest/Response, the jakartaResponse binding,
  Mockito 5, sling-mock 4
- drop the Mockito runner in favor of plain mocks with reflection
  injection, since concurrently running runners collide on Mockito's
  global listener registry under the parallel surefire configuration
- raise the sling-mock resource resolver factory timeout, the default
  500ms is flaky when several Oak-backed contexts start in parallel
- add tests for the previously uncovered classes: SelectorServlet,
  DateUtils, SelectorDetails, BaseFilterFactory, DataFilter and the
  serialization SPI default methods, DefaultDataFilters(Parser),
  DefaultOptionsProcessor, ExcludeDefaultPropertiesProcessor,
  ImportableProcessor, ReferencedProcessor,
  PreventVersionOverrideServletFilter
- fix a latent NPE: SelectorDetails built through the vararg options
  constructor with fewer than two option strings left the options array
  null, crashing SelectorServlet's getOptions().length
- remove an unreachable branch in DateUtils.parseDateTime, parseBest
  either returns one of the two queried types or throws
- remove the module's coverage exemption, restoring the default 1.00
  required instruction coverage ratio
